### ISSUE TITLE
Numba 0.54 only works with numpy>=1.20

### ISSUE THREAD

**Issue by xunkai55** [numba/numba#7339]
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

I have a system using TF 2.6.0 which require `numpy~=1.19`. While if I install Numba (version 0.54, dependent of librosa) at the same time, I got the following runtime errors:

```
RuntimeError: module compiled against API version 0xe but this version of numpy is 0xd
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/tmp/venv/lib/python3.7/site-packages/librosa/__init__.py", line 211, in <module>
    from . import core
  File "/tmp/venv/lib/python3.7/site-packages/librosa/core/__init__.py", line 5, in <module>
    from .convert import *  # pylint: disable=wildcard-import
  File "/tmp/venv/lib/python3.7/site-packages/librosa/core/convert.py", line 7, in <module>
    from . import notation
  File "/tmp/venv/lib/python3.7/site-packages/librosa/core/notation.py", line 8, in <module>
    from ..util.exceptions import ParameterError
  File "/tmp/venv/lib/python3.7/site-packages/librosa/util/__init__.py", line 83, in <module>
    from .utils import *  # pylint: disable=wildcard-import
  File "/tmp/venv/lib/python3.7/site-packages/librosa/util/utils.py", line 10, in <module>
    import numba
  File "/tmp/venv/lib/python3.7/site-packages/numba/__init__.py", line 39, in <module>
    from numba.core.decorators import (cfunc, generated_jit, jit, njit, stencil,
  File "/tmp/venv/lib/python3.7/site-packages/numba/core/decorators.py", line 12, in <module>
    from numba.stencils.stencil import stencil
  File "/tmp/venv/lib/python3.7/site-packages/numba/stencils/stencil.py", line 11, in <module>
    from numba.core import types, typing, utils, ir, config, ir_utils, registry
  File "/tmp/venv/lib/python3.7/site-packages/numba/core/ir_utils.py", line 16, in <module>
    from numba.core.extending import _Intrinsic
  File "/tmp/venv/lib/python3.7/site-packages/numba/core/extending.py", line 19, in <module>
    from numba.core.pythonapi import box, unbox, reflect, NativeValue  # noqa: F401
  File "/tmp/venv/lib/python3.7/site-packages/numba/core/pythonapi.py", line 12, in <module>
    from numba import _helperlib
ImportError: numpy.core.multiarray failed to import
```

If I downgrade it to `numba==0.53`, everything works fine. I guess numba 0.54 wheels are built against a 1.20+ numpy, so it actually requires such a version.

**Comment by xunkai55** [comment]
How to reproduce:

```
pip install numpy==1.19.5
pip install numba==0.54
python -c "import numba"
```

**Comment by xunkai55** [comment]
Resources: looking into numpy code, its C API upgraded from 1.19 to 1.20:

https://github.com/numpy/numpy/blob/main/numpy/core/setup_common.py

So if numba 0.54 is built on API 0xe, it can only runs on those platforms (1.20+).

**Comment by esc** [comment]
@xunkai55 thank you for reporting this. May I ask which Platform (operating system and CPU architecture) you are using? I have tried to reproduce this on OSX (Catalina):

```
$ conda create -n issue7339 python=3.7 pip
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/vhaenel/miniconda3/envs/issue7339

  added / updated specs:
    - pip
    - python=3.7


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    pip-21.0.1                 |   py37hecd8cb5_0         1.8 MB  defaults
    python-3.7.11              |       h88f2d9e_0        18.1 MB  defaults
    wheel-0.37.0               |     pyhd3eb1b0_0          32 KB  defaults
    ------------------------------------------------------------
                                           Total:        19.9 MB

The following NEW packages will be INSTALLED:

  ca-certificates    pkgs/main/osx-64::ca-certificates-2021.7.5-hecd8cb5_1
  certifi            pkgs/main/osx-64::certifi-2021.5.30-py37hecd8cb5_0
  libcxx             pkgs/main/osx-64::libcxx-10.0.0-1
  libffi             pkgs/main/osx-64::libffi-3.3-hb1e8313_2
  ncurses            pkgs/main/osx-64::ncurses-6.2-h0a44026_1
  openssl            pkgs/main/osx-64::openssl-1.1.1k-h9ed2024_0
  pip                pkgs/main/osx-64::pip-21.0.1-py37hecd8cb5_0
  python             pkgs/main/osx-64::python-3.7.11-h88f2d9e_0
  readline           pkgs/main/osx-64::readline-8.1-h9ed2024_0
  setuptools         pkgs/main/osx-64::setuptools-52.0.0-py37hecd8cb5_0
  sqlite             pkgs/main/osx-64::sqlite-3.36.0-hce871da_0
  tk                 pkgs/main/osx-64::tk-8.6.10-hb0a8c7a_0
  wheel              pkgs/main/noarch::wheel-0.37.0-pyhd3eb1b0_0
  xz                 pkgs/main/osx-64::xz-5.2.5-h1de35cc_0
  zlib               pkgs/main/osx-64::zlib-1.2.11-h1de35cc_3



Downloading and Extracting Packages
wheel-0.37.0         | 32 KB     | ############################################################################################################################ | 100%
python-3.7.11        | 18.1 MB   | ############################################################################################################################ | 100%
pip-21.0.1           | 1.8 MB    | ############################################################################################################################ | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate issue7339
#
# To deactivate an active environment, use
#
#     $ conda deactivate

  4.97s user 4.86s system 19% cpu 51.364 total
$ conda activate issue7339

$ pip install numpy==1.19.5
Collecting numpy==1.19.5
  Downloading numpy-1.19.5-cp37-cp37m-macosx_10_9_x86_64.whl (15.6 MB)
     |████████████████████████████████| 15.6 MB 312 kB/s
Installing collected packages: numpy
Successfully installed numpy-1.19.5
$ pip install numba==0.54
Collecting numba==0.54
  Downloading numba-0.54.0-cp37-cp37m-macosx_10_14_x86_64.whl (2.3 MB)
     |████████████████████████████████| 2.3 MB 37 kB/s
Requirement already satisfied: numpy<1.21,>=1.17 in ./miniconda3/envs/issue7339/lib/python3.7/site-packages (from numba==0.54) (1.19.5)
Collecting llvmlite<0.38,>=0.37.0rc1
  Downloading llvmlite-0.37.0-cp37-cp37m-macosx_10_9_x86_64.whl (19.1 MB)
     |████████████████████████████████| 19.1 MB 263 kB/s
Requirement already satisfied: setuptools in ./miniconda3/envs/issue7339/lib/python3.7/site-packages (from numba==0.54) (52.0.0.post20210125)
Installing collected packages: llvmlite, numba
Successfully installed llvmlite-0.37.0 numba-0.54.0
$ python -c "import numba"
0
```

**Comment by xunkai55** [comment]
Thanks for the quick reply! I'm using Ubuntu 18 (Linux x86_64).

**Comment by xunkai55** [comment]
The exact wheel: numba-0.54.0-cp37-cp37m-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (3.4 MB)

**Comment by esc** [comment]
@xunkai55 thank you for following up. We will continue to look into this.

**Comment by gmarkall** [comment]
I can reproduce the issue on Ubuntu 18.04 using the commands given above:

```
pip install numpy==1.19.5
pip install numba==0.54
python -c "import numba"
```

**Comment by esc** [comment]
Note to selves: we need to double check the aarch64 wheels too. They are new with the 0.54.0 release and I strongly suspect they will suffer the same issue.

**Comment by esc** [comment]
Also, we need to make sure that this concerns only the wheels, not the conda packages.

**Comment by sklam** [comment]
I found out why. The buildfarm dag is pinning on np1.15 for py3.7 but Numba `requirements.txt` is asking for np>=1.17. This causes `pip` to fetch the latest.

**Comment by sklam** [comment]
I think this is only affecting all the python3.7 builds where we always pin to np1.15

**Comment by esc** [comment]
@sklam thank you for looking into this! @xunkai55 did you, by any chance, look into other Python versions too?

**Comment by xunkai55** [comment]
@esc Sorry I didn't try other python versions. I indeed only used Python 3.7.

**Comment by esc** [comment]
> @esc Sorry I didn't try other python versions. I indeed only used Python 3.7.

OK, no worries. Note to selves, we should probably look into this.

**Comment by gmarkall** [comment]
There also seems to be the same issue on AArch64 - the manifestation is a little different, but I assume it's the same problem:

```
(base) gmarkall@gm-agx:~$ conda create -n issue-7339 python=3.7
...
#
# To activate this environment, use
#
#     $ conda activate issue-7339
#
# To deactivate an active environment, use
#
#     $ conda deactivate

(base) gmarkall@gm-agx:~$ conda activate issue-7339

(issue-7339) gmarkall@gm-agx:~$ pip install numpy==1.19.5
Collecting numpy==1.19.5
  Downloading numpy-1.19.5-cp37-cp37m-manylinux2014_aarch64.whl (12.4 MB)
     |████████████████████████████████| 12.4 MB 1.2 MB/s 
Installing collected packages: numpy
Successfully installed numpy-1.19.5

(issue-7339) gmarkall@gm-agx:~$ pip install numba==0.54
Collecting numba==0.54
  Downloading numba-0.54.0-cp37-cp37m-manylinux2014_aarch64.manylinux_2_17_aarch64.whl (3.2 MB)
     |████████████████████████████████| 3.2 MB 3.8 MB/s 
Collecting llvmlite<0.38,>=0.37.0rc1
  Downloading llvmlite-0.37.0-cp37-cp37m-manylinux2014_aarch64.whl (24.8 MB)
     |████████████████████████████████| 24.8 MB 164 kB/s 
Requirement already satisfied: numpy<1.21,>=1.17 in /data/gmarkall/miniforge/envs/issue-7339/lib/python3.7/site-packages (from numba==0.54) (1.19.5)
Requirement already satisfied: setuptools in /data/gmarkall/miniforge/envs/issue-7339/lib/python3.7/site-packages (from numba==0.54) (57.4.0)
Installing collected packages: llvmlite, numba
Successfully installed llvmlite-0.37.0 numba-0.54.0

(issue-7339) gmarkall@gm-agx:~$ python -c "import numba"
Illegal instruction (core dumped)
```

**Comment by uschmidt83** [comment]
In case this is useful information, we also observe this issue on Ubuntu with Python 3.7, the combination with other OSes (Windows, macOS) and Python versions (3.6-3.9) works:

https://github.com/stardist/stardist/actions/runs/1159758405

**Comment by sklam** [comment]
New wheels are uploaded. Pip install now should pull the wheel with build number 2:

```
Downloading numba-0.54.0-2-cp37-cp37m-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (3.4 MB)
```

**Comment by esc** [comment]
<img width="812" alt="Screen Shot 2021-09-01 at 10 41 42" src="https://user-images.githubusercontent.com/41563/131640581-a4522885-677a-44bf-a46f-4c650c011b01.png">


Looks good, if others from the community could verify that the CI jobs are running again, we would be happy to close this issue. Thank you again all!

**Comment by uschmidt83** [comment]
Our CI jobs are now running with the updated wheel: https://github.com/stardist/stardist/actions/runs/1183246824

**Comment by stuartarchibald** [comment]
> Our CI jobs are now running with the updated wheel: https://github.com/stardist/stardist/actions/runs/1183246824

@uschmidt83 Many thanks for confirming.

**Comment by PhilipVinc** [comment]
Colab now works, thanks!

**Comment by stuartarchibald** [comment]
> Colab now works, thanks!

Thanks for confirming @PhilipVinc !

**Comment by esc** [comment]
Thank you again all, I would suggest to close this issue now and re-open if need be.

**Comment by stuartarchibald** [comment]
> Thank you again all, I would suggest to close this issue now and re-open if need be.

Agree, thanks all, closing.

**Comment by MacJei** [comment]
Hi guys. We are currently having problems importing the snap library. This may be related to the snap update. Start a new issue?
 python
Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import shap
/opt/conda/lib/python3.7/site-packages/pandas/compat/_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).
  warnings.warn(msg, UserWarning)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.7/site-packages/shap/__init__.py", line 12, in <module>
    from ._explanation import Explanation, Cohorts
  File "/opt/conda/lib/python3.7/site-packages/shap/_explanation.py", line 12, in <module>
    from .utils._general import OpChain
  File "/opt/conda/lib/python3.7/site-packages/shap/utils/__init__.py", line 1, in <module>
    from ._clustering import hclust_ordering, partition_tree, partition_tree_shuffle, delta_minimization_order, hclust
  File "/opt/conda/lib/python3.7/site-packages/shap/utils/_clustering.py", line 4, in <module>
    from numba import jit
  File "/opt/conda/lib/python3.7/site-packages/numba/__init__.py", line 198, in <module>
    _ensure_critical_deps()
  File "/opt/conda/lib/python3.7/site-packages/numba/__init__.py", line 138, in _ensure_critical_deps
    raise ImportError("Numba needs NumPy 1.20 or less")
ImportError: Numba needs NumPy 1.20 or less

numpy==1.21.3
numba==0.54.1
shap==0.40.0

**Comment by esc** [comment]
> Hi guys. We are currently having problems importing the snap library. This may be related to the snap update. Start a new issue? python Python 3.7.3 | packaged by conda-forge | (default, Jul 1 2019, 21:52:21) [GCC 7.3.0] :: Anaconda, Inc. on linux Type "help", "copyright", "credits" or "license" for more information.
> 
> > > > import shap
> > > > /opt/conda/lib/python3.7/site-packages/pandas/compat/_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).
> > > > warnings.warn(msg, UserWarning)
> > > > Traceback (most recent call last):
> > > > File "", line 1, in 
> > > > File "/opt/conda/lib/python3.7/site-packages/shap/**init**.py", line 12, in 
> > > > from ._explanation import Explanation, Cohorts
> > > > File "/opt/conda/lib/python3.7/site-packages/shap/_explanation.py", line 12, in 
> > > > from .utils._general import OpChain
> > > > File "/opt/conda/lib/python3.7/site-packages/shap/utils/**init**.py", line 1, in 
> > > > from ._clustering import hclust_ordering, partition_tree, partition_tree_shuffle, delta_minimization_order, hclust
> > > > File "/opt/conda/lib/python3.7/site-packages/shap/utils/_clustering.py", line 4, in 
> > > > from numba import jit
> > > > File "/opt/conda/lib/python3.7/site-packages/numba/**init**.py", line 198, in 
> > > > _ensure_critical_deps()
> > > > File "/opt/conda/lib/python3.7/site-packages/numba/**init**.py", line 138, in _ensure_critical_deps
> > > > raise ImportError("Numba needs NumPy 1.20 or less")
> > > > ImportError: Numba needs NumPy 1.20 or less
> 
> numpy==1.21.3 numba==0.54.1 shap==0.40.0

@MacJei thank you for asking about this. I don't think there is the need to open a new issue, we already have a pull-request for Numpy 1.21 support here: https://github.com/numba/numba/pull/7483 -- feel free to subscribe to that issue for updates. Numpy 1.21 support will be available with Numba 0.55.

**Comment by MacJei** [comment]
@esc thank you

**Comment by ynsusanto** [comment]
this is what helped me solved the issues on python 3.7:
pip install numpy==1.20.0 --user
pip install numba==0.53 --user

This also help me with import shap error that giving the same things saying numba need numpy 1.20 or less

**Comment by Firestar-Reimu** [comment]
I have installed `numpy=1.21.2` and I cannot use `numba=0.54.1` which will force me to downgrade `numpy to 1.20.3`

What is strange is that `numba=0.53.0` is compatible with `numpy=1.21.2`

**Comment by gmarkall** [comment]
When Numba 0.53 was released, NumPy 1.21 was not available and we were unaware of the issue we'd have with it in the future, so it didn't have an upper bound on the version number.

Numba 0.55 (in RC now: https://numba.discourse.group/t/numba-0-55-0-rc1/1075#numba-0550rc1-announcement-1) will support NumPy 1.21.
